staging/src/k8s.io/client-go/util/jsonpath/parser.go

```diff
@@ -271,12 +271,26 @@ Loop:
 		text = ":"
 	}
 
+	// dict key
+	value := dictKeyRex.FindStringSubmatch(text)
```

Comment:
I don't think reordering this fully fixes the issue

with this setup:
```
kubectl label namespaces foo=a --all --overwrite
kubectl label namespaces bar=b --all --overwrite
kubectl label namespaces baz.qux=c --all --overwrite
```

this is behavior on master:
```
kubectl get namespaces -o jsonpath="{.items[*].metadata.labels['baz.qux']}"
<no output>
kubectl get namespaces -o jsonpath="{.items[*].metadata.labels['foo','bar','baz.qux']}"
a a a a b b b b
```

this is behavior on this PR:
```
kubectl get namespaces -o jsonpath="{.items[*].metadata.labels['baz.qux']}"
c c c c
kubectl get namespaces -o jsonpath="{.items[*].metadata.labels['foo','bar','baz.qux']}"
a a a a b b b b
```

Reply:
I don't think a regex or a strings.Split is sufficient. I think we have to tokenize the value paying attention to whether `,` characters occur inside `'...'` tokens (don't split) or outside them (split)

Reply:
(Note that this is not just reordering the cases, but also changing the implementation of the dict key to just use the value directly rather than parsing it further.)

It's not possible to change how we interpret `labels['foo','bar','baz.qux']` without breaking backward compatibility. That syntax is _the_ way to include a nested field in a union - it's non-standard but documented, and there's no alternate syntax.

The goal of this PR is to stop that behaviour from spilling over to regular key syntax like `labels['baz.qux']`, where anybody who actually wanted to refer to a nested structure would have used some combination of `labels['baz']['qux']` or `labels.baz.qux` anyway.

You're correct that the parsing of unions still does not correctly handle a lot of edge cases (including commas appearing inside quotes). It is at least better with the reordering.

Reply:
> it also makes the behavior of '.' different in case `labels['foo','bar','baz.qux']` than in case `labels['baz.qux']`.

Yes, and indeed that is unfortunate.

> > It's not possible to change how we interpret `labels['foo','bar','baz.qux']` without breaking backward compatibility.
> 
> I think it's fixing a bug, so backwards compatibility is not an issue.

It would be fixing the bug by removing an intentional (and documented) feature without providing a replacement.

I think a better syntax for this feature might have been more like: `labels[@.foo,'bar',@.baz.qux]`. A way forward might be for someone to implement something along those lines, document it, and after a suitable time change the interpretation of the old syntax to be more consistent with what you'd expect from a quoted string.

Reply:
> > It's not possible to change how we interpret labels['foo','bar','baz.qux'] without breaking backward compatibility.
> 
> I think it's fixing a bug, so backwards compatibility is not an issue.

We want to remain as compatible as possible.

Since kube allowed the non-standard `foo.bar` syntax, `foo.bar` is ambiguous whether it means `"foo.bar" attribute of the current object` or `"bar" attribute of the "foo" attribute of the current object`

I think it would be acceptable to do the following with dot-delimited keys:
* See if the current object has an attribute that matches the full dot-delimited key, and use it if it does
* Otherwise fall back to the non-standard object traversal

Reply:
But to do that, we need to actually tokenize the `[...]` content correctly, not use strings.Split